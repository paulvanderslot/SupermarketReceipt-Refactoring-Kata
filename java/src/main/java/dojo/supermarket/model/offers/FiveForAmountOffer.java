package dojo.supermarket.model.offers;

import dojo.supermarket.model.Discount;
import dojo.supermarket.model.Product;

public class FiveForAmountOffer implements SpecificOffer {

    private final double offerPrice;
    private final Product product;

    public FiveForAmountOffer(Product product, double offerPrice) {
        this.offerPrice = offerPrice;
        this.product = product;
    }

    @Override
    public Product getProduct() {
        return product;
    }

    @Override
    public Discount determineDiscount(Product product, double quantity, double unitPrice) {
        double originalPrice = unitPrice * quantity;
        int quantityRoundedOffToBelow = (int) quantity;
        double priceForAllDiscountedProducts = offerPrice * (quantityRoundedOffToBelow / 5);
        double priceForAllNonDiscountedProducts = quantityRoundedOffToBelow % 2 * unitPrice;
        double totalPriceWithDiscount = priceForAllDiscountedProducts + priceForAllNonDiscountedProducts;
        double discountTotal = originalPrice - totalPriceWithDiscount;
        return new Discount(product, getOfferDescription(), -discountTotal);
    }

    private String getOfferDescription() {
        return "5 for" + offerPrice;
    }
}
