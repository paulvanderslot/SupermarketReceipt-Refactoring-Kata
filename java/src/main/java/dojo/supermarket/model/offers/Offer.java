package dojo.supermarket.model.offers;

import dojo.supermarket.model.Discount;
import dojo.supermarket.model.Product;
import dojo.supermarket.model.ProductQuantity;

public interface Offer {

    Product getProduct();

    Discount determineDiscount(ProductQuantity productQuantity, double unitPrice);

    boolean doesApplyFor(ProductQuantity productQuantity);
}

