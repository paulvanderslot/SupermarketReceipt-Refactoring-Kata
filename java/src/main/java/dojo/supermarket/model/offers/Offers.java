package dojo.supermarket.model.offers;

import dojo.supermarket.model.ProductQuantity;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Optional;
import java.util.function.Predicate;
import java.util.stream.Stream;

public class Offers {

    private final Collection<Offer> offers = new ArrayList<>();

    public void add(Offer offer) {
        Optional<Offer> offerForSameProduct = offers.stream().filter(offerForSameProduct(offer)).findAny();
        offerForSameProduct.ifPresent(offers::remove);
        offers.add(offer);
    }

    private static Predicate<Offer> offerForSameProduct(Offer offer) {
        return it -> it.getProduct().equals(offer.getProduct());
    }

    public Stream<Offer> stream() {
        return offers.stream();
    }

    public Optional<Offer> findMatchingOffer(ProductQuantity productQuantity) {
        return offers.stream().filter(it -> it.doesApplyFor(productQuantity)).findAny();

    }
}
