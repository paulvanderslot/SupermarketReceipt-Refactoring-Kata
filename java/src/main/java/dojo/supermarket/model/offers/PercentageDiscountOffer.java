package dojo.supermarket.model.offers;

import dojo.supermarket.model.Discount;
import dojo.supermarket.model.Product;

public class PercentageDiscountOffer implements Offer {

    private static final String OFFER_DESCRIPTION = "3 for 2";
    private final Product product;
    private final int percentageDiscount;

    public PercentageDiscountOffer(Product product, int percentageDiscount) {
        this.product = product;
        this.percentageDiscount = percentageDiscount;
    }

    @Override
    public Product getProduct() {
        return product;
    }

    @Override
    public Discount determineDiscount(Product product, double quantity, double unitPrice) {
        double originalPrice = unitPrice * quantity;
        double discountAmount = originalPrice * percentageDiscount / 100;
        return new Discount(product, getOfferDescription(), -discountAmount);
    }

    private String getOfferDescription() {
        return percentageDiscount + "% off";
    }

}
