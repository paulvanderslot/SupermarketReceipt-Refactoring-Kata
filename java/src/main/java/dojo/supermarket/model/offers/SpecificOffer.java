package dojo.supermarket.model.offers;

import dojo.supermarket.model.Discount;
import dojo.supermarket.model.Product;

public interface SpecificOffer {

    Product getProduct();

    Discount determineDiscount(Product product, double quantity, double unitPrice);
}

