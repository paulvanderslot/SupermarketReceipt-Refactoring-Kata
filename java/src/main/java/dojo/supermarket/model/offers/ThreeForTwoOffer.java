package dojo.supermarket.model.offers;

import dojo.supermarket.model.Discount;
import dojo.supermarket.model.Product;

public class ThreeForTwoOffer implements Offer {

    private static final String OFFER_DESCRIPTION = "3 for 2";
    private final Product product;

    public ThreeForTwoOffer(Product product) {
        this.product = product;
    }

    @Override
    public Product getProduct() {
        return product;
    }

    @Override
    public Discount determineDiscount(Product product, double quantity, double unitPrice) {
        if (quantity < 3) return null;

        double originalPrice = unitPrice * quantity;
        int timesTheOfferApplies = (int) quantity / 3;

        double offerPrice = 2 * unitPrice;
        double discountAmount = originalPrice - ((timesTheOfferApplies * offerPrice) + (int) quantity % 3 * unitPrice);
        return new Discount(product, OFFER_DESCRIPTION, -discountAmount);
    }

}
