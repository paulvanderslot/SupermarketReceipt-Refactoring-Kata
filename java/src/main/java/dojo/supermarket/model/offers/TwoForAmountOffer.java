package dojo.supermarket.model.offers;

import dojo.supermarket.model.Discount;
import dojo.supermarket.model.Product;

public class TwoForAmountOffer implements Offer {

    private final double offerPrice;
    private final Product product;

    public TwoForAmountOffer(Product product, double offerPrice) {
        this.offerPrice = offerPrice;
        this.product = product;
    }

    @Override
    public Product getProduct() {
        return product;
    }

    @Override
    public Discount determineDiscount(Product product, double quantity, double unitPrice) {
        if (quantity < 2) return null;

        double originalPrice = unitPrice * quantity;
        int quantityRoundedOffToBelow = (int) quantity;
        double priceForAllDiscountedProducts = offerPrice * (quantityRoundedOffToBelow / 2);
        double priceForAllNonDiscountedProducts = quantityRoundedOffToBelow % 2 * unitPrice;
        double totalPriceWithDiscount = priceForAllDiscountedProducts + priceForAllNonDiscountedProducts;
        double amountOfDiscount = originalPrice - totalPriceWithDiscount;
        return new Discount(product, getOfferDescription(), -amountOfDiscount);
    }

    private String getOfferDescription() {
        return "2 for " + offerPrice;
    }
}
