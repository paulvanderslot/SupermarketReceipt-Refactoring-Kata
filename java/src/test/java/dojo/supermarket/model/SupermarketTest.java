package dojo.supermarket.model;

import dojo.supermarket.ReceiptPrinter;
import dojo.supermarket.model.offers.FiveForAmountOffer;
import dojo.supermarket.model.offers.PercentageDiscountOffer;
import dojo.supermarket.model.offers.ThreeForTwoOffer;
import dojo.supermarket.model.offers.TwoForAmountOffer;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class SupermarketTest {
    private SupermarketCatalog catalog;
    private final Product toothbrush = new Product("toothbrush", ProductUnit.EACH);
    private final Product apples = new Product("apples", ProductUnit.KILO);
    private final double applePrice = 1.99;
    private final double toothbrushPrice = 0.99;

    @BeforeEach
    void beforeEach() {
        catalog = new FakeCatalog();
        catalog.addProduct(toothbrush, toothbrushPrice);
        catalog.addProduct(apples, applePrice);
    }

    @Test
    void shouldGiveNoDiscountsWhenNoneApply() {
        Teller teller = new Teller(catalog);
        teller.addSpecialOffer(new PercentageDiscountOffer(toothbrush, 10));

        ShoppingCart cart = new ShoppingCart();
        double appleQuantity = 2.5;
        cart.addItemQuantity(apples, appleQuantity);

        Approvals.verify(new ReceiptPrinter().printReceipt(teller.checksOutArticlesFrom(cart)));

    }

    @Test
    void shouldApply10percentDiscount() {
        Teller teller = new Teller(catalog);
        int discountPercentage = 10;
        teller.addSpecialOffer(new PercentageDiscountOffer(toothbrush, discountPercentage));

        ShoppingCart cart = new ShoppingCart();
        double toothbrushQuantity = 3;
        cart.addItemQuantity(toothbrush, toothbrushQuantity);

        Approvals.verify(new ReceiptPrinter().printReceipt(teller.checksOutArticlesFrom(cart)));
    }

    @Test
    void shouldApply20percentDiscount() {
        Teller teller = new Teller(catalog);
        int discountPercentage = 20;
        teller.addSpecialOffer(new PercentageDiscountOffer(toothbrush, discountPercentage));

        ShoppingCart cart = new ShoppingCart();
        double toothbrushQuantity = 3;
        cart.addItemQuantity(toothbrush, toothbrushQuantity);

        Approvals.verify(new ReceiptPrinter().printReceipt(teller.checksOutArticlesFrom(cart)));
    }

    @Test
    void shouldApply2ForAmountWhenTwoAreBought() {
        Teller teller = new Teller(catalog);
        double twoForAmountPrice = 1.50;
        teller.addSpecialOffer(new TwoForAmountOffer(toothbrush, twoForAmountPrice));

        ShoppingCart cart = new ShoppingCart();
        double toothbrushQuantity = 2;
        cart.addItemQuantity(toothbrush, toothbrushQuantity);

        Approvals.verify(new ReceiptPrinter().printReceipt(teller.checksOutArticlesFrom(cart)));
    }

    @Test
    void shouldApply2ForAmountWhen3AreBought() {
        Teller teller = new Teller(catalog);
        double twoForAmountPrice = 1.50;
        teller.addSpecialOffer(new TwoForAmountOffer(toothbrush, twoForAmountPrice));

        ShoppingCart cart = new ShoppingCart();
        double toothbrushQuantity = 3;
        cart.addItemQuantity(toothbrush, toothbrushQuantity);

        Approvals.verify(new ReceiptPrinter().printReceipt(teller.checksOutArticlesFrom(cart)));
    }

    @Test
    void shouldApply2ForAmountWhen4AreBought() {
        Teller teller = new Teller(catalog);
        double twoForAmountPrice = 1.50;
        teller.addSpecialOffer(new TwoForAmountOffer(toothbrush, twoForAmountPrice));

        ShoppingCart cart = new ShoppingCart();
        double toothbrushQuantity = 4;
        cart.addItemQuantity(toothbrush, toothbrushQuantity);

        Approvals.verify(new ReceiptPrinter().printReceipt(teller.checksOutArticlesFrom(cart)));
    }


    @Test
    void shouldApply3For2When3AreBought() {
        Teller teller = new Teller(catalog);
        teller.addSpecialOffer(new ThreeForTwoOffer(apples));

        ShoppingCart cart = new ShoppingCart();
        double boughtQuantity = 3;
        cart.addItemQuantity(apples, boughtQuantity);

        Approvals.verify(new ReceiptPrinter().printReceipt(teller.checksOutArticlesFrom(cart)));
    }

    @Test
    void shouldApply3For2When4AreBought() {
        Teller teller = new Teller(catalog);
        teller.addSpecialOffer(new ThreeForTwoOffer(apples));

        ShoppingCart cart = new ShoppingCart();
        double boughtQuantity = 4;
        cart.addItemQuantity(apples, boughtQuantity);

        Approvals.verify(new ReceiptPrinter().printReceipt(teller.checksOutArticlesFrom(cart)));
    }

    @Test
    void shouldNotApply5ForAmountWhen4AreBought() {
        Teller teller = new Teller(catalog);
        teller.addSpecialOffer(new FiveForAmountOffer(apples, 6.00));

        ShoppingCart cart = new ShoppingCart();
        double boughtQuantity = 4;
        cart.addItemQuantity(apples, boughtQuantity);

        Approvals.verify(new ReceiptPrinter().printReceipt(teller.checksOutArticlesFrom(cart)));
    }

    @Test
    void shouldApply5ForAmountWhen5AreBought() {
        Teller teller = new Teller(catalog);
        teller.addSpecialOffer(new FiveForAmountOffer(apples, 6.00));

        ShoppingCart cart = new ShoppingCart();
        cart.addItemQuantity(apples, 5);

        Approvals.verify(new ReceiptPrinter().printReceipt(teller.checksOutArticlesFrom(cart)));
    }

    @Test
    void shouldApply5ForAmountWhen6AreBoughtIn2Goes() {
        Teller teller = new Teller(catalog);
        teller.addSpecialOffer(new FiveForAmountOffer(apples, 6.00));

        ShoppingCart cart = new ShoppingCart();
        cart.addItemQuantity(apples, 3);
        cart.addItemQuantity(apples, 3);


        Approvals.verify(new ReceiptPrinter().printReceipt(teller.checksOutArticlesFrom(cart)));
    }

    @Test
    void shouldApply5ForAmountWhen10AreBought() {
        Teller teller = new Teller(catalog);
        teller.addSpecialOffer(new FiveForAmountOffer(apples, 6.00));

        ShoppingCart cart = new ShoppingCart();
        cart.addItemQuantity(apples, 10);

        Approvals.verify(new ReceiptPrinter().printReceipt(teller.checksOutArticlesFrom(cart)));
    }

    @Test
    void shouldApplyOnlyOneDiscountOnOneItem() {
        Teller teller = new Teller(catalog);
        teller.addSpecialOffer(new PercentageDiscountOffer(apples, 50));
        teller.addSpecialOffer(new ThreeForTwoOffer(apples));

        ShoppingCart cart = new ShoppingCart();
        double boughtQuantity = 4;
        cart.addItemQuantity(apples, boughtQuantity);

        Approvals.verify(new ReceiptPrinter().printReceipt(teller.checksOutArticlesFrom(cart)));
    }

    @Test
    void shouldApplyMultipleDiscountsOnMultipleItems() {
        Teller teller = new Teller(catalog);
        teller.addSpecialOffer(new PercentageDiscountOffer(apples, 50));
        teller.addSpecialOffer(new FiveForAmountOffer(toothbrush, 2.00));

        ShoppingCart cart = new ShoppingCart();
        cart.addItemQuantity(apples, 4);
        cart.addItemQuantity(toothbrush, 6);


        Approvals.verify(new ReceiptPrinter().printReceipt(teller.checksOutArticlesFrom(cart)));
    }


}
