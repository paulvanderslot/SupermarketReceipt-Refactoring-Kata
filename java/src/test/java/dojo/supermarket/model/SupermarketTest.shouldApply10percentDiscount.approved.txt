toothbrush                          2.97
  0.99 * 3
10% off(toothbrush)                -0.30

Total:                              2.67
